stingray/bexvar.py

```diff
@@ -0,0 +1,241 @@
+        m0 = -1
+    # highest count rate (including background)
+    c = scipy.special.gammaincinv(src_counts + 1, 0.999) / rate_conversion
+    m1 = np.log10(c.max())
+
+    # add a bit of padding to the bottom and top
+    lo = m0 - 0.05 * (m1 - m0)
+    hi = m1 + 0.05 * (m1 - m0)
+    span = hi - lo
+    if lo < -1:
+        log_src_crs_grid = np.linspace(-1.0, hi, int(np.ceil(density_gp * (hi + 1.0))))
+    else:
+        log_src_crs_grid = np.linspace(lo, hi, int(np.ceil(density_gp * 1.05 * span)))
+
+    return log_src_crs_grid
+
+
+def _estimate_source_cr_marginalised(log_src_crs_grid, src_counts, bkg_counts, bkg_area, \
+        rate_conversion):
+    """
+    Compute the PDF at positions in log(source count rate)s grid ``log_src_crs_grid``
+    for observing ``src_counts`` counts in the source region of size src_area,
+    and ``bkg_counts`` counts in the background region of size ``bkg_area``.
+
+    Parameters
+    ----------
+    log_src_crs_grid : iterable, `:class:numpy.array` of floats
+        An array (grid) of log(Source count rates).
+    src_counts : float
+        Source region counts in one of the bin.
+    bkg_counts : float
+        Background region counts in the bin.
+    bkg_area : float
+        Background region area in the bin.
+    rate_conversion : float
+        Rate conversion constant for the bin. Used to convert counts to count rate.
+
+    Returns
+    -------
+    weights: float
+    """
+    # background counts give background count rates deterministically
+    N = 1000
+    u = np.linspace(0, 1, N)[1:-1]
+    bkg_cr = scipy.special.gammaincinv(bkg_counts + 1, u) / bkg_area
+
+    def prob(log_src_cr):
+        src_cr = 10**log_src_cr * rate_conversion
+        like = scipy.stats.poisson.pmf(src_counts, src_cr + bkg_cr).mean()
+        return like
+
+    weights = np.array([prob(log_src_cr) for log_src_cr in log_src_crs_grid])
+
+    if weights.sum() <= 0:
+        print("Sum is" , weights.sum(), \
+             "\n range of log_src_count_rate_grid ",  log_src_crs_grid[0], log_src_crs_grid[-1])
+        warnings.warn("Weight problem! sum is <= 0")
+
+    weights /= weights.sum()
+
+    return weights
+
+def _calculate_bexvar(log_src_crs_grid, pdfs):
+    """
+    Assumes that the source count rate is log-normal distributed.
+    returns posterior samples of the mean and standard deviation of that distribution.
+
+    Parameters
+    ----------
+    log_src_crs_grid : Iterable, `:class:numpy.array` of flots
+        An array (grid) of log(Source count rates).
+    pdfs : Iterable, `:class:List` of floats
+        An array of PDFs for each object defined over the
+        log(source count rate) grid ``log_src_crs_grid``.
+
+    Returns
+    -------
+    log_sigma : iterable, `:class:numpy.array` of flots.
+        An array of posterior samples of log(standard deviation) or that of
+        log(Bayesian excess varience).
+    """
+
+    if not can_sample:

```

Comment:
You can setup a test that, when ultranest is not installed, calls this function and catches the ImportError.

Reply:
[Test](https://github.com/mihirtripathi97/stingray/blob/GSoC-22/stingray/tests/test_bexvar.py#:~:text=def%20test_ultranest_not_installed(self)%3A) added for this.